File: kielo_shared/vocab/content_bridge_surface_type.py
"""kielo_shared.vocab.content_bridge_surface_type — Python mirror of
kielo-shared/vocab/contentbridgesurfacetype.go.

Content Bridge Arc 1 (2026-06-07): cross-language SoT for the 4
canonical surface-type wire-string values used by the Bridge's
reader endpoint at::

    GET /internal/api/v3/content-bridge/items/{item_id}/surfaces
        ?surface_type=article,video_caption,scenario,exercise_prompt

Mirrors the Go-side typed alias + constants + iteration container at
the wire-string level.

## Architectural shape

  - ``ContentBridgeSurfaceType`` typed alias (str newtype) matching
    Go's ``vocab.ContentBridgeSurfaceType``
  - 4 ``Final[ContentBridgeSurfaceType]`` constants — byte-equivalent
    values to Go SoT
  - ``ALL_CONTENT_BRIDGE_SURFACE_TYPES`` ``FrozenSet`` iteration
    container
  - ``is_valid_content_bridge_surface_type(s)`` validator helper

## Cross-language parity

The contract test at
``tests/contract/content_bridge_surface_type_vocabulary_contract_test.go``
asserts every Go constant has a matching Python constant AND every
Python constant has a matching Go constant. Adding a new value in one
language without the other fails the gate.

## Arc 1 readiness

  - ``SURFACE_ARTICLE`` + ``SURFACE_VIDEO_CAPTION`` — populated
    end-to-end via existing ``cms_<lang>.occurrences`` writes.

  - ``SURFACE_SCENARIO`` + ``SURFACE_EXERCISE_PROMPT`` — empty
    junction tables in Arc 1; readers return ``[]`` until Arcs 2+3
    wire producers.

## Python consumer call sites (Arc 1)

  - kielolearn-engine/.../services/new_item_encountered_notification_author.py
    consumes the Bridge via the Python client (kielo_shared.content_bridge.client)
    to resolve the deep_link destination per the Sweep-MMM
    no-fallback discipline (silence is the right answer when no
    surface resolves).
"""

from __future__ import annotations

from typing import Final, FrozenSet, NewType

# Typed alias mirroring Go-side vocab.ContentBridgeSurfaceType.
ContentBridgeSurfaceType = NewType("ContentBridgeSurfaceType", str)


# Production-ready surface types (2 — populated end-to-end in Arc 1).

SURFACE_ARTICLE: Final[ContentBridgeSurfaceType] = ContentBridgeSurfaceType("article")
"""News article paragraphs. Reader reads ``cms_<lang>.occurrences``
WHERE ``source_type='article'``. Surface_id is the content_version_id;
``paragraph_id`` + ``sentence_text`` + ``original_token_phrase`` carry
the grounding snippet."""

SURFACE_VIDEO_CAPTION: Final[ContentBridgeSurfaceType] = ContentBridgeSurfaceType(
    "video_caption"
)
"""KTV video caption cues. Reader reads ``cms_<lang>.occurrences``
WHERE ``source_type='video'``. Surface_id is the content_version_id;
``caption_index`` + ``timestamp_start`` + ``sentence_text`` carry the
grounding snippet + seek target."""


# Producer-pending surface types (2 — empty junction tables in Arc 1;
# readers return [] until producers wire up).

SURFACE_SCENARIO: Final[ContentBridgeSurfaceType] = ContentBridgeSurfaceType("scenario")
"""Convo scenario turns + hints. Empty in Arc 1; producer wired in
Arc 3 (sub-recon-driven choice: post-session correction extraction
OR LLM batch annotation OR author-time annotation)."""

SURFACE_EXERCISE_PROMPT: Final[ContentBridgeSurfaceType] = ContentBridgeSurfaceType(
    "exercise_prompt"
)
"""kielolearn-engine generated exercise prompts. Empty in Arc 1;
producer wired in Arc 2 (LLM emit-time annotation at
structured_output.py persist point)."""


# Iteration container — frozenset of all 4 canonical values. Used by:
#   - The Go↔Python parity contract test.
#   - The producer-scan contract test.
#   - Operator tooling that needs to enumerate every surface type.
ALL_CONTENT_BRIDGE_SURFACE_TYPES: FrozenSet[ContentBridgeSurfaceType] = frozenset(
    {
        SURFACE_ARTICLE,
        SURFACE_VIDEO_CAPTION,
        SURFACE_SCENARIO,
        SURFACE_EXERCISE_PROMPT,
    }
)


def is_valid_content_bridge_surface_type(value: str) -> bool:
    """Return True when value is one of the canonical wire strings.

    Use this at API request-validation layers (e.g. when constructing
    a Bridge client query) to reject typo'd inputs at the application
    boundary before they hit the reader endpoint's HTTP 400.
    """
    return ContentBridgeSurfaceType(value) in ALL_CONTENT_BRIDGE_SURFACE_TYPES


__all__ = [
    "ContentBridgeSurfaceType",
    "SURFACE_ARTICLE",
    "SURFACE_VIDEO_CAPTION",
    "SURFACE_SCENARIO",
    "SURFACE_EXERCISE_PROMPT",
    "ALL_CONTENT_BRIDGE_SURFACE_TYPES",
    "is_valid_content_bridge_surface_type",
]

File: vocab/contentbridgesurfacetype.go
// Package vocab — Content Bridge Arc 1 (2026-06-07)
//
// ContentBridgeSurfaceType typed-vocab SoT for the Content Bridge
// cross-feature reverse-index. 15th instance of the typed-vocab SoT
// pattern (alongside achievement.go, itemtype.go, scenariosourcetype.go,
// pushkey.go, emailsubjectkey.go, kielo-shared/userlang/sources.go,
// kielo-shared/userlang/grantsource.go, kielo-shared/errors,
// kielo-shared/events/{cmswriterendpoint,outboxeventtype,publisheventtype,
// behavioral,useraction,notificationcreated}.go).
//
// Background:
//
// The Content Bridge answers the question "given (item_type, item_id,
// user_id, language), where else has this learner encountered this
// word or grammar concept across Kielo's content surfaces?"
//
// Four surface types span the answer space. Each is owned by a
// different service:
//
//   - "article"          — kielo-content-service reads from
//     cms_<lang>.occurrences (source_type='article').
//     Populated by kielo-ingest-processor at
//     article-publish time. Production-ready;
//     6,575 rows in cms_fi as of Arc 1 ship.
//
//   - "video_caption"    — kielo-content-service reads from
//     cms_<lang>.occurrences (source_type='video').
//     Populated by kielo-ingest-processor video
//     adapter at KTV-processed time. Note: a
//     parallel writer also populates
//     cms_<lang>.kielotv_mindmaps.graph; canonical
//     source for the Bridge is occurrences (see
//     ADR §5 Drive-by + §9 Open question 2).
//
//   - "scenario"         — kielo-convo reads from a NEW per-language
//     junction (planned: convo.scenario_item_references
//     OR users_conversations.scenario_item_references
//     pending the convo schema rename). Empty in
//     Arc 1; producer wires in Arc 3.
//
//   - "exercise_prompt"  — kielolearn-engine reads from a NEW per-lang
//     junction klearn_<lang>.exercise_item_references.
//     Empty in Arc 1; producer wires in Arc 2.
//
// Arc-1 contract: the reader endpoint at
// GET /internal/api/v3/content-bridge/items/{item_id}/surfaces
// accepts a comma-separated ?surface_type=... filter taking these wire
// strings. Empty results for scenario + exercise_prompt are expected
// until Arcs 2+3 populate.
//
// Adding a new ContentBridgeSurfaceType value:
//
//  1. Add the constant here with the canonical wire string.
//  2. Extend AllContentBridgeSurfaceTypes for iteration.
//  3. Add a sibling constant to kielo_shared/vocab/content_bridge_surface_type.py.
//  4. Wire the producer + reader for the new surface.
//  5. Update the contract tests' required-set in
//     tests/contract/content_bridge_surface_type_vocabulary_contract_test.go.
//
// The 3 contract tests enforce parity:
//
//	(a) Producer-side closed-set scan ✓ TestContentBridgeSurfaceTypeNoStaleLiteralsInProducer
//	(b) Go↔Python parity              ✓ TestContentBridgeSurfaceTypeGoPythonParity
//	(c) SoT cardinality + canonical   ✓ TestContentBridgeSurfaceTypeSoTNonEmpty
//
// (No DB CHECK constraint mirror — the surface type is a query-param
// vocabulary, not a column value. The closed-set discipline lives at
// the application boundary via IsKnownContentBridgeSurfaceType.)
package vocab

// ContentBridgeSurfaceType is the typed alias for the canonical
// content-surface vocabulary used by the Bridge's wire shape
// (query param + response field).
//
// Wire format is a string so request URLs and JSON responses stay
// stable across schema migrations; this Go type narrows the
// producer-side surface (reader endpoint, client SDK, contract test)
// so a stray literal can't be silently accepted.
type ContentBridgeSurfaceType string

// String returns the wire string. Allows the typed constant to be
// used directly in URL building, log fields, and JSON marshaling
// without an explicit conversion at every call site.
func (s ContentBridgeSurfaceType) String() string {
	return string(s)
}

// Production-ready surface types (2 — populated end-to-end in Arc 1).
const (
	// SurfaceArticle: news article paragraphs. Reader reads
	// cms_<lang>.occurrences WHERE source_type='article'. The
	// surface_id is the content_version_id; paragraph_id +
	// sentence_text + original_token_phrase carry the grounding
	// snippet.
	SurfaceArticle ContentBridgeSurfaceType = "article"

	// SurfaceVideoCaption: KTV video caption cues. Reader reads
	// cms_<lang>.occurrences WHERE source_type='video'. The
	// surface_id is the content_version_id; caption_index +
	// timestamp_start + sentence_text carry the grounding snippet
	// + seek target.
	SurfaceVideoCaption ContentBridgeSurfaceType = "video_caption"
)

// Producer-pending surface types (2 — empty junction tables in Arc 1;
// readers return [] until producers wire up).
const (
	// SurfaceScenario: convo scenario turns + hints. Empty in
	// Arc 1; producer wired in Arc 3 (sub-recon-driven choice:
	// post-session correction extraction OR LLM batch annotation
	// OR author-time annotation).
	SurfaceScenario ContentBridgeSurfaceType = "scenario"

	// SurfaceExercisePrompt: kielolearn-engine generated exercise
	// prompts. Empty in Arc 1; producer wired in Arc 2 (LLM
	// emit-time annotation at structured_output.py persist point).
	SurfaceExercisePrompt ContentBridgeSurfaceType = "exercise_prompt"
)

// AllContentBridgeSurfaceTypes is the canonical iteration order.
// Used by:
//   - The reader endpoint's default surface filter (when no
//     ?surface_type=... param is supplied).
//   - The Go↔Python parity contract test (asserts every Python
//     mirror constant has a Go sibling and vice-versa).
//   - The producer-scan contract test (asserts no production code
//     emits a wire string outside this set).
var AllContentBridgeSurfaceTypes = []ContentBridgeSurfaceType{
	SurfaceArticle,
	SurfaceVideoCaption,
	SurfaceScenario,
	SurfaceExercisePrompt,
}

// IsKnownContentBridgeSurfaceType returns true when s is one of the
// canonical values. Used by the reader endpoint to reject typo'd
// query-param values at the application boundary (HTTP 400) before
// they reach the SELECT planner.
func IsKnownContentBridgeSurfaceType(s ContentBridgeSurfaceType) bool {
	for _, known := range AllContentBridgeSurfaceTypes {
		if known == s {
			return true
		}
	}
	return false
}
